test/integration/network/dual_stack_test.go

```diff
@@ -0,0 +1,336 @@
+			opts.ServiceClusterIPRanges = "10.0.0.0/16,2001:db8:1::/48"
+		},
+		ModifyServerConfig: func(config *master.Config) {
+		},
+	})
+
+	ipFamily := v1.IPv4Protocol
+	svc := createService(&ipFamily, v1.ServiceTypeClusterIP)
+	// Create a service setting the IPFamily and check that it's created with the correct IP family
+	if _, err := clientSet.CoreV1().Services(metav1.NamespaceDefault).Create(svc); err != nil {
+		t.Errorf("unexpected error text: %v", err)
+	}
+	svc = createService(nil, v1.ServiceTypeClusterIP)
+	svc.ObjectMeta.Name = "test2"
+	// Create a service without IPFamily and check that it's created with the default IP family
+	if _, err := clientSet.CoreV1().Services(metav1.NamespaceDefault).Create(svc); err != nil {
+		t.Errorf("unexpected error text: %v", err)
+	}
+
+	svcs, err := clientSet.CoreV1().Services(metav1.NamespaceAll).List(metav1.ListOptions{})
+	if err != nil {
+		t.Fatalf("unexpected success, and error getting the services: %v", err)
+	}
+	for _, s := range svcs.Items {
+		if err = validateServiceAndClusterIPFamily(&s, ipFamily); err != nil {
+			t.Fatalf("unexpected service IP family %s: %v", s.Spec.ClusterIP, err)
+		}
+	}
+}
+
+// TestDualStackIPv6DefaultCluster tests a dual stack API server with IPv6 as default family
+func TestDualStackIPv6DefaultCluster(t *testing.T) {
+	defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.IPv6DualStack, true)()
+
+	stopCh := make(chan struct{})
+	defer close(stopCh)
+	clientSet, _ := framework.StartTestServer(t, stopCh, framework.TestServerSetup{
+		ModifyServerRunOptions: func(opts *options.ServerRunOptions) {
+			opts.ServiceClusterIPRanges = "2001:db8:1::/48,10.0.0.0/16"
+		},
+		ModifyServerConfig: func(config *master.Config) {
+		},
+	})
+
+	ipFamily := v1.IPv6Protocol
+	svc := createService(&ipFamily, v1.ServiceTypeClusterIP)
+	// Create a service setting the IPFamily and check that it's created with the correct IP family
+	if _, err := clientSet.CoreV1().Services(metav1.NamespaceDefault).Create(svc); err != nil {
+		t.Errorf("unexpected error text: %v", err)
+	}
+	svc = createService(nil, v1.ServiceTypeClusterIP)
+	svc.ObjectMeta.Name = "test2"
+	// Create a service without IPFamily and check that it's created with the default IP family
+	if _, err := clientSet.CoreV1().Services(metav1.NamespaceDefault).Create(svc); err != nil {
+		t.Errorf("unexpected error text: %v", err)
+	}
+
+	svcs, err := clientSet.CoreV1().Services(metav1.NamespaceAll).List(metav1.ListOptions{})
+	if err != nil {
+		t.Fatalf("unexpected success, and error getting the services: %v", err)
+	}
+	for _, s := range svcs.Items {
+		if err = validateServiceAndClusterIPFamily(&s, ipFamily); err != nil {
+			t.Fatalf("unexpected service IP family %s: %v", s.Spec.ClusterIP, err)
+		}
+	}
+
+}
+
+// TestDualStackSkew tests two dual stack API server with and without DualStack enabled
+func TestDualStackSkew(t *testing.T) {
```

Comment:
@liggitt  I'll need help here to understand how to "create" this testing
https://github.com/kubernetes/kubernetes/issues/81035